text/trace/0170-sampling-probability.md

```diff
@@ -0,0 +1,677 @@
+system describes the use of sampling to control the cost of trace
+collection at scale.  Dapper's early Sampler algorithm, referred to as
+an "inflationary" approach (although not published in the paper), is
+reproduced here.
+
+This kind of Sampler allows non-root spans in a trace to raise the
+probability of tracing, using a conditional probability formula shown
+below.  Traces produced in this way are complete sub-trees, not
+necessarily complete.  This technique is successful especially in
+systems where a high-throughput service on occasion calls a
+low-throughput service.  Low-throughput services are meant to inflate
+their sampling probability.
+
+The use of this technique requires propagating the head inclusion
+probability (as discussed for the `Parent` sampler) of the incoming
+Context and whether it was sampled, in order to calculate the
+probability of starting to sample a new "sub-root" in the trace.
+
+Using standard notation for conditional probability, `P(x)` indicates
+the probability of `x` being true, and `P(x|y)` indicates the
+probability of `x` being true given that `y` is true.  The axioms of
+probability establish that:
+
+```
+P(x)=P(x|y)*P(y)+P(x|not y)*P(not y)
+```
+
+The variables are:
+
+- **`H`**: The head inclusion probability of the parent context that
+  is in effect, independent of whether the parent context was sampled
+- **`I`**: The inflationary sampling probability for the span being
+  started.
+- **`D`**: The decision probability for whether to start a new sub-root.
+
+This Sampler cannot lower sampling probability, so if the new span is
+started with `H >= I` or when the context is already sampled, no new
+sampling decisions are made.  If the incoming context is already
+sampled, the adjusted count of the new span is `1/H`.
+
+Assuming `H < I` and the incoming context was not sampled, we have the
+following probability equations:
+
+```
+P(span sampled) = I
+P(parent sampled) = H
+P(span sampled | parent sampled) = 1
+P(span sampled | parent not sampled) = D
+```
+
+Using the formula above, 
+
+```
+I = 1*H + D*(1-H)
+```
+
+solve for D:
+
+```
+D = (I - H) / (1 - H)
+```
+
+Now the Sampler makes a decision with probability `D`.  Whether the
+decision is true or false, propagate `I` as the new head inclusion
+probability.  If the decision is true, begin recording a sub-rooted
+trace with adjusted count `1/I`.
+
+## Proposed specification text
+
+The following text will be added to the semantic conventions for
+recording the Sampler name and adjusted count (if known) as
+OpenTelemetry Span attributes.
+
+```
+# Semantic conventions for Sampled spans
+
+This document defines how to describe an what sampling was performed
+when recording a span that has had sampling logic applied.
+
+Span sampling attributes support computing metrics about spans that
+are part of a sampled trace from knowing their sampling inclusion
+probability.
+
+The _adjusted count_ of a span is defined as follows:
```

Comment:
I think this is addressed in https://github.com/open-telemetry/oteps/pull/170#discussion_r678474697. The natural count of the Span is 1, which should be used when there is no adjustment.

Reply:
I think "natural count" is somewhat subjective. For sites that use traditional probabilistic head-based sampling with propagation of sampling decision, the "natural count" for spans would be 1/p. These sites would also be more sensitive (due to cost) to including additional span attributes than the sites that can "afford" 100% sampling (e.g. Jaeger SDKs only record sampling params on the root span). I am weary of making a choice in the spec that would penalize one group at the expense of the other. We may be forced to make such a choice (unless we provide for alternative means of capturing the info), in which case it would make sense to poll many users who would be affected by it.

Reply:
> I think "natural count" is somewhat subjective.

You are calling an adjusted count natural. I feel that is incorrect. There is only one event corresponding with each Span, thus the natural event count is 1.

Your position, if I understand it, is that the cost of reporting adjusted count when it is not 1, which supports span-to-metrics pipelines, is not worth the expense of having always on. I am open to other's input on this.

I would like to question the argument that a system that adopts probability sampling because 100% of spans is too expensive will object to the incremental cost of reporting adjusted count.  A slight reduction of the sampling rate will compensate for the additional cost of supporting span-to-metrics. I am interested in knowing more about how existing systems compute metrics about spans if they are forced to assemble traces before generating those metrics. Do any such span-to-metrics pipelines exist for Jaeger systems?

Reply:
> I am interested in knowing more about how existing systems compute metrics about spans if they are forced to assemble traces before generating those metrics.

What's the problem with assembling traces first? It introduces an ingestion delay, but then the actual insights one gets from these "metrics" are coupled with availability of traces anyway.

Reply:
This is about more than ingestion delay. We have to add the root span's duration to the ingestion delay before metrics can be derived for the entire trace. If the root span never completes or is dropped, an entire trace worth of spans can't be counted.

You could also ask: why not just use metrics in the first place? Answer 1: Spans contain logs, we get information about the population of log events from a these span metrics. Answer 2: This data stream supports high-cardinality metric pipelines.

This was covered in the narrative above, especially lines 436-440 and 463-469. "Spans are countable" is a requirement I am standing behind, because real-time span-to-metrics requires it.

Reply:
I'm resolving this comment because I believe I have address the question. 

> insights one gets from these "metrics" are coupled with availability of traces anyway.

This could be a source of confusion. The goal when performing span-to-metrics is to learn about the population of spans in their own right. At Lightstep, when we see an uninteresting span we might just count it an move on. 

I've completed a prototype of this code and OTEP #168 by now, and will update #168 with more detail about propagation. @yurishkuro please let me know if there is something more I can explain or address about the span data model conventions in this PR, and I will update #168 tomorrow. Thanks.